Clean up and simplify the treatment of implicit binding (#532)
ghc-proposals/ghc-proposals: proposals/0448-type-variable-scoping.rst

This component of this proposal is taken
from the not-yet-accepted proposal `#238`_, changing the name of what I now call
``-XExtendedForAllScope``.

``-XImplicitBinds`` is the combination of ``-XImplicitBindsForAll`` and ``-XPatternSignatureBinds``) from accepted
proposal `#285`_, and differ from that proposal in that:

  - The extensions are combined for fewer knobs when the motivations are the same.

  - ``RULES`` was forgotten as an example.

>   - Examples of implicit binds falsely categorized as pattern signature binds are now explicitly included.

Reviewer: Concretely, this means that in `data A (a :: k)`, `k` is now in the same category as kind signature occurrences?

Author: ```suggestion
  - Examples of implicit binds falsely categorized as pattern signature binds are now properly included under the proposed change specification.
```

  It thus seems necessary to separate out the problematic ``-XExtendedForAllScope``
  from the other components of ``-XScopedTypeVariables``.

A secondary goal is to clean up some issues with proposal `#285`_ while simplifying things:

* ``-XImplicitForAll`` and ``-XPatternSignatureBinds`` have the exact same
  motivation, and it is unclear why one would ever want one without the other.

* ``-XImplicitForAll`` and ``-XPatternSignatureBinds`` *missed* some of the cases in the examples, which clearly are implicit binding forms meant to be turned off per the overall motivation, but nonetheless slipped through the cracks of the drafting process.

Having separated out ``-XExtendedForAllScope``, it seemed strange to have a ``-XRumpEndOfOldScopedTypeVariables``
extension, and so I've introduced separate ``-XMethodTypeVariables`` and ``-XPatternSignatures``.

> ``-XPatternSignatures`` is more narrow than it's GHC 6.x namesake so as not to violate any principles by default, and to match ``-XKindSignatures`` which is narrow in the same fashion.

Reviewer: ```suggestion
``-XPatternSignatures`` is more narrow than its GHC 6.x namesake so as not to violate any principles by default, and to match ``-XKindSignatures`` which is narrow in the same fashion.
```

What I really want in this case is a pattern signature,
and it would be nice if I could just state that ``PatternSignatures``.

Motivation for not doing binding
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

``-XPatternSignatures`` as proposed here is more narrow than its GHC 6 namesake.
This is because it just allows signatures using already-bound variables, and doesn't include any implicit binding mechanism for variables in the signature that aren't yet bound.
(That instead is left for ``-XImplicitBinds``.)

We here discuss the motivation for this decision.

Consistency
"""""""""""

This more narrow formulation of ``-XPatternSignatures`` matches ``KindSignatures``.
``KindSignatures`` doesn't allow implicit binds for a rather roundabout reason: implicit binds would imply implicit kind-level foralls, which would require ``-XPolyKinds``.
Given the other extensions being proposed here, we can retroactively reinterpret this as a simple syntactic rule: ``-XKindSignatures`` alone doesn't do implicit binding.

> Now, both extension just allow, respectively, term-level and type-level signatures, with no other functionality like implicit binding mechanisms also thrown in.

Author: Done!

Conservativism for standardization
""""""""""""""""""""""""""""""""""

With both of these signatures being very minimal, I think they would be easy uncontroversial candidates for a new language report.

> Conversely, all implicit binding constructs are very fraught with a complicated mix of upsides and downsides, we and should only standardize them with great care.

Author: It is possible I duplicated something by mistake. I will double check, thanks.

What I really want in this case is a pattern signature, and it would be nice if I could
just state that ``PatternSignatures``.

Motivation for not doing binding
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Consistency
"""""""""""

Firstly, it matches ``KindSignatures``.
``KindSignatures`` doesn't allow implicit binds for a rather roundabout reason: implicit binds would imply implicit kind-level foralls, which would require ``-XPolyKinds``.
Given the other extensions being proposed here, we can retroactively reinterpret this as a simple syntactic rule: ``-XKindSignatures`` alone doesn't do implicit binding.

> Now, ``PatternSignature``

Reviewer: You appear to have abandoned this sentence

By banning implicit binding, we side-step that difference.

After all, given::

  data Foo (a :: k)

desugars to::

  data Foo @k (a :: k)

a new Haskeller might conceivably think::

  \(Foo (a :: k) -> ..

desugars to::

  \(Foo @k (a :: k) ->

or::

  \ @k (Foo (a :: k) ->

which happen to be true in some simple common cases, but are in fact incorrect in general.

That it takes a complicated example to show why these false desugarings aren't true in general make this is a huge educational stumbling block!

Unified Namespace
^^^^^^^^^^^^^^^^^

If `#270`_ is accepted, there will be a way to program Haskell with "morally" one namespace for types and terms alike.
However, there is one exception to the unification of namespaces: lower case variables in type signatures bound "like terms" still are treated as free and implicitly bound instead::

  t = Int
  x :: t -- sugar for 'forall t. t', not 't ~ Int'
  x = 0

  t = Int
  foo (x :: t) = 0 -- sugar for 'foo = let t = _ in \(x :: t) -> 0'

> Should the ``t`` in ``x :: t`` cause implicit ``forall t.`` and ``let t = I in`` to be synthesized or not?

Author: Actually it shouldn't be, but I agree it was unclear. Hopefully better now.

Unified Namespace
^^^^^^^^^^^^^^^^^

If `#270`_ is accepted, there will be a way to program Haskell with "morally" one namespace for types and terms alike.
However, there is one exception to the unification of namespaces: lower case variables in type signatures bound "like terms" still are treated as free and implicitly bound instead::

  t = Int
  x :: t -- sugar for 'forall t. t', not a use of 't' resolving to 'Int'
  x = 0

  t = Int
  foo (x :: t) = 0 -- sugar for 'foo = let t = _ in \(x :: t) -> 0'

Should the ``t`` in each ``x :: t`` cause implicit ``forall t.`` and ``let t = _ in`` to be synthesized or not?

With ``-XNoImplicitBinds``, we know it will not, and thus can refer to the ``t`` defined above, once such a reference is possible (left to another proposal).

Without ``-XImplicitBinds`` we have no choice but do the implicit desugaring that violates the unified namespace abstraction
--- otherwise we would be *forklike* in the sense that rather than giving new meaning to previously invalid programs, we would be *changing* the meaning of previously *valid* programs.

> Put another way, ``-XImplicitBinds`` and any ``-XUnifiedNamespace`` must be mutually exclusive, so we need a way for Haskell programmers to opt out of ``-XImplicitBinds`` first.

Author: The issue is that we start with an invalid program, and we make it valid in two *different* ways, one with liberalized namespace lookup and the "term 2 type translate", the other with implicit binding, but there is no monotonic way to reconcile those things. Each extension is non-fork-like in isolation, but for the combination we are stuck.

> Put another way, ``-XImplicitBinds`` and any ``-XUnifiedNamespace`` must be mutually exclusive, so we need a way for Haskell programmers to opt out of ``-XImplicitBinds`` first.

Reviewer 2: Ah, this is quite interesting. Can you make this more explicit and detailed in the proposal? It's not obvious to me yet what's happening.

> Put another way, ``-XImplicitBinds`` and any ``-XUnifiedNamespace`` must be mutually exclusive, so we need a way for Haskell programmers to opt out of ``-XImplicitBinds`` first.

Author: I expanded upon it in https://github.com/Ericson2314/ghc-proposals/blob/type-variables/proposals/0448-type-variable-scoping.rst#unified-namespacing-and-extension-monotonicity

@Ericson2314 doesn't wish to argue whether these choices do or don't actually help learning, but just state that some people have opinions that they do and there is no technical reason GHC cannot accommodate them.

Consistency
^^^^^^^^^^^

Notice how today that out-of-scope variables in negative position signatures are implicitly bound in *different* ways depending on whether they are type variables (in pattern signatures) or kind variables (in negative position kind signatures).
By banning implicit binding, we side-step that difference.

After all, given::

  data Foo (a :: k)

desugars to::

  data Foo @k (a :: k)

a new Haskeller might conceivably think::

  \(Foo (a :: k) -> ..

desugars to::

  \(Foo @k (a :: k) ->

or::

  \ @k (Foo (a :: k) ->

which happen to be true in some simple common cases, but are in fact incorrect in general.

That it takes a complicated example to show why these false desugarings aren't true in general make this is a huge educational stumbling block!

Unified Namespace
^^^^^^^^^^^^^^^^^

If `#270`_ is accepted, there will be a way to program Haskell with "morally" one namespace for types and terms alike.

> However, there is one exception to the unification of namespaces: lower case variables in type signatures bound "like terms" still are treated as free and implicitly bound instead::

Reviewer 2: .

Author: Ah this is also an artifact of the bad diffing I am afraid; there is a pattern signature binds examples added below which does *not* desugar to a `forall`. I therefore had to reword this to describe both examples at once.

  t = Int

>   x :: t -- sugar for 'forall t. t', not a use of 't' resolving to 'Int'

Reviewer 2: .

Author: Perhaps I was being too liberal modifying this because...it is in fact my own writing taking verbatim from #285 here that I am nit-picking!

The issue that is that `t` is *not* sugar for `t ~ Int`, because `t` is a type and `t ~ Int` is a constraint. I was in #285 trying to express that `t` is bound as `Int` but did so in what strikes me in hindsight as a sloppy manner.

I suppose I could do this in a separate PR or not at all. Let me know what you prefer.

  x = 0

  t = Int
  foo (x :: t) = 0 -- sugar for 'foo = let t = _ in \(x :: t) -> 0'

> Should the ``t`` in each ``x :: t`` cause implicit ``forall t.`` and ``let t = _ in`` to be synthesized or not?

Reviewer 2: .

Author: The sentence that appears removed is on the line below.

Without ``-XImplicitBinds`` we have no choice but do the implicit desugaring that violates the unified namespace abstraction
--- otherwise we would be *forklike* in the sense that rather than giving new meaning to previously invalid programs, we would be *changing* the meaning of previously *valid* programs.
Put another way, ``-XImplicitBinds`` and any ``-XUnifiedNamespace`` must be mutually exclusive, so we need a way for Haskell programmers to opt out of ``-XImplicitBinds`` first.

Proposed Change Specification
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Points below up to and including the new (backward-compatible) definition of
``-XScopedTypeVariables`` come from not-yet-accepted proposal `#238`_.
``-XImplicitBinds`` is a fixed and simplified (via combining extensions) version of accepted proposal `#285`_.

1. Re-purpose deprecated extension ``-XPatternSignatures``. With ``-XPatternSignatures``, we
   allow type signatures in patterns. These signatures can mention in-scope
   type variables as variable occurrences, but can not bind type variables without the

>    separate ``-XImplicitBinds`` extension. Do note that extension is on by default, however.

Reviewer 2: .

Author: Isn't this more than a rewording? Before, implicit pattern signature binds were gated by `-XPatternSignatures` alone, but warned. Now, they are gated by the combination of `-XPatternSignatures` and `-XImplicitBinds`.

Yes, the difference in practice is subtle because the latter is on by default, but generally when I write these things I try to specify the effects of extensions and the defaulting of extensions separately. (Dealing with both at once I find confusing.)

      #. Types in a ``deriving`` clause.
         Example:
         ``data T deriving (C a)``
         becomes
         ``data T deriving (forall a. C a)``.

      #. Instance heads, including standalone-deriving instances.
         Example:
         ``instance Show a => Show (Maybe a)``
         becomes
         ``instance forall a. Show a => Show (Maybe a)``.

      #. Type and data family instances, as well as closed type family equations.
         Example:
         ``type instance F (Maybe a) = Int``
         becomes
         ``type instance forall a. F (Maybe a) = Int``.

      #. ``RULES`` pragmas.
         Example: ``{-# RULES "name" forall (x :: Maybe a). foo x = 5 #-}``
         becomes ``{-# RULES "name" forall a. forall (x :: Maybe a). foo x = 5 #-}``.
         (The double-\ ``forall`` syntax separates type variables like ``a`` from
         term variables like ``x``.)

      This is the former ``-XImplicitForAll`` from accepted but unimplemented proposal `#285`_;
      the only change is including ``RULES`` pragmas, which @Ericson2314 simply forgot to include in `#285`_ (his own admission).

   #. Implicit binds in pattern signatures:

>       Out-of-scope type variables written in a pattern signature would be bound there and would remain in scope over the same region of code that term-level variables introduced in a pattern scope over.

Reviewer 2: This is lacking an example.

      This is the former ``-XPatternSignatureBinds`` from accepted, unimplemented proposal `#285`_.

   #. Implicit binds in kind signatures:

>       Out-of-scope type variables written in a negative position kind signature (positive ones are implicit foralls) are bound as implicit capital lambdas to the left of the parameter they occur in.

Reviewer 2: This is lacking an example as well.

Alternatives
~~~~~~~~~~~~

1. We could further break down ``-XImplicitBinds``, like before.

>    But fixing the drafting error would require a *third* extension, ``-XNegativeSignatureBinds``, in addition to the original two.

Reviewer 2: I think that you are unambiguously allowed to propose modification to unimplemented proposal.

Author: My point wasn't that I felt that I couldn't change it, but that I didn't want to be responsible for for *even more* language extensions. 448 already adds a lot already, and I didn't want to in modifying the proposal give credence to that objection to 448.

Reviewer 2: This is not what you're saying though. As it currently stands, I think it's rather misleading in fact.

Author: Ah I now gather that "But fixing the drafting error" is what is confusing. See https://github.com/ghc-proposals/ghc-proposals/pull/532/files#r1038961339 where I tried to reword this.

Not-yet-accepted amendment `#291`_ says that type variables scope just like term variables: they can be shadowed.
Accordingly, ``f (Just @a x) = ...`` would always, unconditionally bind a new type variable ``a``, possibly shadowing any in-scope type variable ``a``.
This design supports the `Visibility Orthogonality Principle`_, which states that the presence of an ``@`` should affect only whether a thing is visible or not, not other characteristics (like its shadowing and scoping behavior).

> Additionally, this choice edges us closer to the `Lexical Scoping Principle`_, because we no longer have to check whether ``a`` is in scope before identifying the ``a`` in ``f (Just @a x) = ...`` is a binding site or an occurrence.

Reviewer 2: This is no change here, correct?

Author: Ah just one word. It was referring to the "Local Lexical Scoping Principle" which was never merged.

Alternatives
~~~~~~~~~~~~

> 1. We could further break down ``-XImplicitBinds``, like before.
> 
>    But fixing the drafting error would require a *third* extension, ``-XNegativeSignatureBinds``, in addition to the original two.
>    This would allow more conservative defaults --- we must have Haskell98 implicit foralls but not the others which are all guarded behind language extensions today.
> 
>    However, @Ericson2314 sensed there is a weariness with too many extensions coming from this, and so didn't do it.

Author: ```suggestion
1. We could further break down ``-XImplicitBinds``, like before.

   This would allow more conservative defaults --- we must have Haskell98 implicit foralls but not the others which are all guarded behind language extensions today.
   
   But fixing the drafting error in `#285`_, where certain examples meant to be including under `-XPatternSignatureBinds` that wouldn't be because they are not "pattern signatures", in this manner would require a *third* extension, ``-XNegativeSignatureBinds``, in addition to the original two.
   @Ericson2314 sensed there is a weariness with so many extensions coming from this proposal (448) already, and so didn't want to pursue this route.
```

   But fixing the drafting error would require a *third* extension, ``-XNegativeSignatureBinds``, in addition to the original two.
   This would allow more conservative defaults --- we must have Haskell98 implicit foralls but not the others which are all guarded behind language extensions today.

   However, @Ericson2314 sensed there is a weariness with too many extensions coming from this, and so didn't do it.

Costs and Drawbacks
~~~~~~~~~~~~~~~~~~~

Unified Namespacing and extension monotonicity
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Unified namespacing was touted as a beneficiary of ``-XNoImplicitBinds`` above.
But on the other hand, `270#`_ and `281#_`, the latter of which is accepted and partially implemented, adopt a model where variables in types resolving to variables defined in the term namespace as a fallback unconditionally.
This is indeed backwards compatible, however it breaks the property of ``-XImplicitBinds`` being strictly *non-forklike* in allowing only more programs, not changing the meaning of existing programs.

To wit, if ::

  t = Int
  x :: t -- out of scope, no type variable `t` in scope.
  x = 0

> is an invalid program, we can *either* give make it valid by saying the second ``t`` is a use or implicit bind, but we cannot do *both*.

Reviewer 2: I assume you meant
```suggestion
is an invalid program, we can *either* make it valid by saying the second ``t`` is a use or implicit bind, but we cannot do *both*.
```

   (Right now, pattern synonyms require all such quantifications to occur before any term arguments,
   but accepted proposal `#402`_ allows these quantifications to occur in any order in data constructors.)

#. In accordance with the `Visibility Orthogonality Principle`_,
   the rules that determine whether a variable occurrence is a binding site or a use site are not affected by the presence of a ``@``.
   At the time of writing, the rules are as follows:

   * Outside pattern signatures,
     variable occurrences are considered binding sites,
     shadowing any other in-scope variables.
     It is an error to bring the same type variable into scope in two (or more) places within the same match group.

   * Inside pattern signatures (i.e. on the right-hand side of ``pat :: sig``),
     occurrences of in-scope type variables are usages,
     whereas occurrences of out-of-scope type variables create implicit bindings.

>      It is allowed to mention the same out-of-scope variable more than once.

Reviewer 2: This is unrelated to the current principal, isn't it?

Author: Oh whoops, I accidentally pushed my merging of https://github.com/ghc-proposals/ghc-proposals/pull/556 into this to fix comments.

Author: Forced pushed to remove merge.

   it does not trigger any behavior associated with partial type signatures.
   In particular, ``-XPartialTypeSignatures`` is not necessary, and no diagnostic is produced.

#. As with term variables, it is an error to bring the same type variable into scope in two (or more) places within the same pattern.

> 

Reviewer 2: Why did you remove this?

Author: Oh whoops, I accidentally pushed my merging of https://github.com/ghc-proposals/ghc-proposals/pull/556 into this to fix comments.

That one deleted this because it is not quite correct and the expanded above portion now covers it.

because we no longer have to check whether ``a`` is in scope before identifying the ``a`` in ``f (Just @a x) = ...`` is a binding site or an occurrence.

The other change in this restatement is the use of new extension ``-XTypeAbstractions`` instead of the current status of piggy-backing on the combination of ``-XTypeApplications`` and ``-XScopedTypeVariables`` (*both* need to be enabled today).

> This proposal suggests instead that ``-XScopedTypeVariables`` and ``-XScopedTypeVariables`` should enable this feature in a deprecated manner, to encourage users to use ``-XTypeAbstractions`` instead.

Reviewer 3: ```suggestion
This proposal suggests instead that ``-XScopedTypeVariables`` and ``-XTypeApppications`` should enable this feature in a deprecated manner, to encourage users to use ``-XTypeAbstractions`` instead.
```

Author: @nomeata Sorry for the noise, the other PR will take care of that. This one is on hold until it and one I have yet to open are resolved -- trying to avoid huge diffs and get the least controversial parts taken care of first.

I just pushed to this one to make sure I wasn't accumulating conflicts.

Reviewer 3: This is a different PR? Ah, right! Ignore me then :-)

The effects of this proposal are written out in the individual sections.
Here, I summarize the effects on the principles_.

#. The `Explicit Variable Principle`_ is made to hold, by allowing explicit binders for type variables for existentials and the variables bound by an inner ``forall`` in a higher-rank
   type.

#. The `Lexical Scoping Principle`_ is made to hold with ``-XNoImplicitBinds``

>    Indeed, the purpose of ``-XNoImplicitBinds`` is to be the single extension which is both necessary and sufficient to do this.

Reviewer 2: Well, not completely: the Template Haskell objection still holds I think.

Author: ```suggestion

#. The `Lexical Scoping Principle`_ outside of Template Haskell is made to hold with ``-XNoImplicitBinds``

   Indeed, the purpose of ``-XNoImplicitBinds`` is to be the single extension which is both necessary and sufficient to do this.
   (The Template Haskell issue is something would be solved by more rigorous notations of hygiene. That has little to do with ``-XNoImplicitBinds`` as currently specified, and is much more work of a very different nature.) 
```

   The `Lexical Scoping Principle`_, part (a), is upheld.
   Binders occur in patterns, after ``forall``, in
   ``let`` declarations, and a few other discrete places in the AST -- and
   nowhere else. In particular, binders do not occur in pattern signatures.

>    The `Lexical Scoping Principle`_ part (a) is made to hold,

Reviewer 2: ```suggestion
   The `Lexical Scoping Principle`_ part (b) is made to hold,
```

   type.

#. The `Lexical Scoping Principle`_ is made to hold with ``-XNoImplicitBinds``

   Indeed, the purpose of ``-XNoImplicitBinds`` is to be the single extension which is both necessary and sufficient to do this.

   The `Lexical Scoping Principle`_, part (a), is upheld.
   Binders occur in patterns, after ``forall``, in
   ``let`` declarations, and a few other discrete places in the AST -- and
   nowhere else. In particular, binders do not occur in pattern signatures.

   The `Lexical Scoping Principle`_ part (a) is made to hold, by describing pattern-signature binds as occurrences
   and making type applications in patterns unconditionally bring new variables
   into scope.

   This would not be the case with the treatment of in-scope variables as originally written
   in `#126`_, where the choice between a binding site and an occurrence depends on whether a
   type variable is in scope.

#. The `Syntactic Unification Principle`_ is bolstersted by ``-XNoImplicitBinds``

   As discussed in the "Consistency" section of the motivation for that extension, the different forms of implicit binding we have today work quite differently.

>    In many case, those different forms are chiefly distinguished by being confined to one of the type- or term- level, e.g. regular pattern for type synonym right hand sides.

Author: Rewrote and split. Hopefully better now.

   A lack of inconsistency on this front therefore is part of the type and term syntax unification espoused by this principle.

#. The `Explicit Binding Principle`_ is made to hold under ``-XNoImplicitBinds`` and ``-XPatternSignatures`` by side-stepping the need for new explicit syntax.

   Making ``-XPatternSignatures`` not imply implicit bindings keeps that exntension in accordance with the `Explicit Binding Principle`_.
   That principle says implicit binding constructions should have explicit counterparts they desugar to.

   In general, solutions to `Lexical Scoping Principle`_ are also solutions to `Explicit Binding Principle`_.

>    It is just for implicit forms ones wishes to leave enabled that explicit syntax is needed, and explicit syntax forms an additional solution to the `Explicit Binding Principle`_ (and not the `Lexical Scoping Principle`_).

Reviewer: ```suggestion
   It is just for implicit forms one wishes to leave enabled that explicit syntax is needed, and explicit syntax forms an additional solution to the `Explicit Binding Principle`_ (and not the `Lexical Scoping Principle`_).
```

Author: ```suggestion
   It is just for implicit forms which one wishes to leave enabled that explicit syntax is needed, and explicit syntax forms an additional solution to the `Explicit Binding Principle`_ (and not the `Lexical Scoping Principle`_).
```
